src/main/java/org/apache/sysds/runtime/matrix/data/LibCommonsMath.java

```diff
@@ -209,6 +218,339 @@ private static MatrixBlock computeSolve(MatrixBlock in1, MatrixBlock in2) {
 		
 		return DataConverter.convertToMatrixBlock(solutionMatrix);
 	}
+
+	/**
+	 * Computes the eigen decomposition of a symmetric matrix.
+	*
+	* @param in The input matrix to compute the eigen decomposition on.
+	* @param threads The number of threads to use for computation.
+	* @return An array of MatrixBlock objects containing the real eigen values and eigen vectors.
+	*/
+	public static MatrixBlock[] computeEigenDecompositionSymm(MatrixBlock in, int threads) {
+ 
+		// TODO: Verify matrix is symmetric
+		final double[] mainDiag = new double[in.rlen];
+		final double[] secDiag = new double[in.rlen - 1];
+
+		final MatrixBlock householderVectors = transformToTridiagonal(in, mainDiag, secDiag, threads);
+
+		// TODO: Consider using sparse blocks
+		final double[] hv = householderVectors.getDenseBlockValues();
+		MatrixBlock houseHolderMatrix = getQ(hv, mainDiag, secDiag);
+
+		MatrixBlock[] evResult = findEigenVectors(mainDiag, secDiag, houseHolderMatrix, EIGEN_MAX_ITER, threads);
+
+		MatrixBlock realEigenValues = evResult[0];
+		MatrixBlock eigenVectors = evResult[1];
+
+		realEigenValues.setNonZeros(realEigenValues.rlen * realEigenValues.rlen);
+		eigenVectors.setNonZeros(eigenVectors.rlen * eigenVectors.clen);
+
+		return new MatrixBlock[] { realEigenValues, eigenVectors };
+	}
+
+	private static MatrixBlock transformToTridiagonal(MatrixBlock matrix, double[] main, double[] secondary, int threads) {
+		final int m = matrix.rlen;
+
+		// MatrixBlock householderVectors = ;
+		MatrixBlock householderVectors = matrix.extractTriangular(new MatrixBlock(m, m, false), false, true, true);
+		if (householderVectors.isInSparseFormat()) {
+			householderVectors.sparseToDense(threads);      
+		}
+
+		final double[] z = new double[m];
```

Comment:
This temporary allocation seems like the only thing that limits you from parallelizing the entire for loop.